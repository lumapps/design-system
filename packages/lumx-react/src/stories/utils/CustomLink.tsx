/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
