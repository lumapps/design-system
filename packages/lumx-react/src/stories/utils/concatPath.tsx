export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
