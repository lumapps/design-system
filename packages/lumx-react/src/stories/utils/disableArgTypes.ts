export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
