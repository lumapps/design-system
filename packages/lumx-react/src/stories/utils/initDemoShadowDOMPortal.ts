/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
