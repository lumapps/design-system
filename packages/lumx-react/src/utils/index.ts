/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
