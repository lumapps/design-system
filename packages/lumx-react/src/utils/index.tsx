export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
