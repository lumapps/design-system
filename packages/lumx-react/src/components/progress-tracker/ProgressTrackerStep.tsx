import React, { FocusEventHandler, forwardRef, KeyboardEventHandler, useCallback } from 'react';

import classNames from 'classnames';

import { Icon, InputHelper, InputLabel, Kind, Size } from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { mdiAlertCircle, mdiCheckCircle, mdiRadioboxBlank, mdiRadioboxMarked } from '@lumx/icons';
import { useTabProviderContext } from '../tabs/state';

/**
 * Defines the props of the component.
 */
export interface ProgressTrackerStepProps extends GenericProps {
    /** Children are not supported. */
    children?: never;
    /** Whether the step should be in error state or not. */
    hasError?: boolean;
    /** Helper text. */
    helper?: string | null;
    /** Native id property. */
    id?: string;
    /** Whether the current step is active or not. */
    isActive?: boolean;
    /** Whether the current step is completed or not. */
    isComplete?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Label text. */
    label?: string | null;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'ProgressTrackerStep';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ProgressTrackerStepProps> = {};

/**
 * ProgressTrackerStep component.
 *
 * Implements WAI-ARIA `tab` role {@see https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html#rps_label}
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const ProgressTrackerStep: Comp<ProgressTrackerStepProps, HTMLButtonElement> = forwardRef((props, ref) => {
    const {
        className,
        disabled,
        hasError,
        helper,
        id,
        isActive: propIsActive,
        isComplete,
        isDisabled = disabled,
        label,
        onFocus,
        onKeyPress,
        tabIndex = -1,
        ...forwardedProps
    } = props;
    const state = useTabProviderContext('tab', id);
    const isActive = propIsActive || state?.isActive;

    const changeToCurrentTab = useCallback(() => {
        if (isDisabled) {
            return;
        }
        state?.changeToTab();
    }, [isDisabled, state]);

    const handleFocus: FocusEventHandler = useCallback(
        (event) => {
            onFocus?.(event);
            if (state?.shouldActivateOnFocus) {
                changeToCurrentTab();
            }
        },
        [changeToCurrentTab, onFocus, state?.shouldActivateOnFocus],
    );

    const handleKeyPress: KeyboardEventHandler = useCallback(
        (event) => {
            onKeyPress?.(event);
            if (event.key !== 'Enter') {
                return;
            }
            changeToCurrentTab();
        },
        [changeToCurrentTab, onKeyPress],
    );

    const getIcon = (): string => {
        if (isComplete) {
            return mdiCheckCircle;
        }

        if (isActive) {
            return hasError ? mdiAlertCircle : mdiRadioboxMarked;
        }

        return mdiRadioboxBlank;
    };

    return (
        <button
            ref={ref}
            {...forwardedProps}
            type="button"
            id={state?.tabId}
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                    hasError,
                    isActive,
                    isClickable: state && !isDisabled,
                    isComplete,
                }),
            )}
            onClick={changeToCurrentTab}
            onKeyPress={handleKeyPress}
            onFocus={handleFocus}
            role="tab"
            tabIndex={isActive ? 0 : tabIndex}
            aria-disabled={isDisabled}
            aria-selected={isActive}
            aria-controls={state?.tabPanelId}
        >
            <Icon className={`${CLASSNAME}__state`} icon={getIcon()} size={Size.s} />

            <InputLabel htmlFor={state?.tabId || ''} className={`${CLASSNAME}__label`}>
                {label}
            </InputLabel>

            {helper && (
                <InputHelper kind={hasError ? Kind.error : Kind.info} className={`${CLASSNAME}__helper`}>
                    {helper}
                </InputHelper>
            )}
        </button>
    );
});
ProgressTrackerStep.displayName = COMPONENT_NAME;
ProgressTrackerStep.className = CLASSNAME;
ProgressTrackerStep.defaultProps = DEFAULT_PROPS;
