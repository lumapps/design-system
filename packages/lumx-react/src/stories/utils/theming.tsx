import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
