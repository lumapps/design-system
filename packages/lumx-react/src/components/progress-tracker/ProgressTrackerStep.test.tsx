import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import React, { ReactElement } from 'react';

import { setupTabProviderMocks } from '../tabs/test.mocks';
import { ProgressTrackerStep, ProgressTrackerStepProps } from './ProgressTrackerStep';

const CLASSNAME = ProgressTrackerStep.className as string;

// Mock useTabProviderContext.
jest.mock('../tabs/state', () => {
    const state = jest.requireActual('../tabs/state');
    return { ...state, useTabProviderContext: jest.fn(), useTabProviderContextState: jest.fn() };
});

type SetupProps = Partial<ProgressTrackerStepProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: ProgressTrackerStepProps = { label: 'Test Step Label', ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<ProgressTrackerStep {...props} />);

    return { props, wrapper };
};

describe(`<${ProgressTrackerStep.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render', () => {
            setupTabProviderMocks({ index: 1 });
            const { wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });

        it('should render active state', () => {
            setupTabProviderMocks({ index: 3, isActive: true });
            const { wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
        });

        it('should render complete state', () => {
            setupTabProviderMocks({ index: 3, isComplete: true });
            const { wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
        });

        it('should render erroneous state', () => {
            setupTabProviderMocks({ index: 3, hasError: true });
            const { wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
        });

        it('should render active erroneous state', () => {
            setupTabProviderMocks({ index: 3, isActive: true, hasError: true });
            const { wrapper } = setup({});
            expect(wrapper).toMatchSnapshot();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use the given props to add classes', () => {
            setupTabProviderMocks();
            const props: any = {
                isActive: true,
                hasError: true,
                isComplete: true,
            };
            const { wrapper } = setup(props);

            Object.keys(props).forEach((prop) => {
                expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: prop, value: props[prop] }));
            });
        });
    });

    // 3. Test events.
    describe('Events', () => {
        it('should trigger `onChange` when focused if the provider is configured to do so', () => {
            const { changeToTab } = setupTabProviderMocks({ shouldActivateOnFocus: true });
            const { wrapper } = setup({}, false);

            wrapper.simulate('focus');
            expect(changeToTab).toHaveBeenCalledTimes(1);
        });

        it('should trigger `onChange` when clicked', () => {
            const { changeToTab } = setupTabProviderMocks();
            const { wrapper } = setup({}, false);

            wrapper.simulate('click');
            expect(changeToTab).toHaveBeenCalledTimes(1);
        });

        it('should trigger `onChange` when pressing `enter`', () => {
            const { changeToTab } = setupTabProviderMocks();
            const { wrapper } = setup({}, false);

            wrapper.simulate('keypress', { key: 'Enter' });
            expect(changeToTab).toHaveBeenCalledTimes(1);
        });

        it('should not trigger `onChange` when pressing any other key', () => {
            const { changeToTab } = setupTabProviderMocks();
            const { wrapper } = setup({}, false);

            wrapper.simulate('keypress', { key: 'a' });
            expect(changeToTab).not.toHaveBeenCalled();
        });
    });

    // Common tests suite.
    commonTestsSuite(
        (...args: any[]) => {
            setupTabProviderMocks();
            return setup(...args);
        },
        { className: 'wrapper', prop: 'wrapper' },
        { className: CLASSNAME },
    );
});
