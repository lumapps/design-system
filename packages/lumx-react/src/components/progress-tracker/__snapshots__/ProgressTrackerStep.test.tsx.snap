// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressTrackerStep> Snapshots and structure should render 1`] = `
<button
  aria-controls="tab-panel-1"
  aria-selected={false}
  className="lumx-progress-tracker-step lumx-progress-tracker-step--is-clickable"
  id="tab-1"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={-1}
  type="button"
>
  <Icon
    className="lumx-progress-tracker-step__state"
    icon="M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
    size="s"
  />
  <InputLabel
    className="lumx-progress-tracker-step__label"
    htmlFor="tab-1"
    theme="light"
  >
    Test Step Label
  </InputLabel>
</button>
`;

exports[`<ProgressTrackerStep> Snapshots and structure should render active erroneous state 1`] = `
<button
  aria-controls="tab-panel-3"
  aria-selected={true}
  className="lumx-progress-tracker-step lumx-progress-tracker-step--is-active lumx-progress-tracker-step--is-clickable"
  id="tab-3"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={0}
  type="button"
>
  <Icon
    className="lumx-progress-tracker-step__state"
    icon="M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,7A5,5 0 0,0 7,12A5,5 0 0,0 12,17A5,5 0 0,0 17,12A5,5 0 0,0 12,7Z"
    size="s"
  />
  <InputLabel
    className="lumx-progress-tracker-step__label"
    htmlFor="tab-3"
    theme="light"
  >
    Test Step Label
  </InputLabel>
</button>
`;

exports[`<ProgressTrackerStep> Snapshots and structure should render active state 1`] = `
<button
  aria-controls="tab-panel-3"
  aria-selected={true}
  className="lumx-progress-tracker-step lumx-progress-tracker-step--is-active lumx-progress-tracker-step--is-clickable"
  id="tab-3"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={0}
  type="button"
>
  <Icon
    className="lumx-progress-tracker-step__state"
    icon="M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,7A5,5 0 0,0 7,12A5,5 0 0,0 12,17A5,5 0 0,0 17,12A5,5 0 0,0 12,7Z"
    size="s"
  />
  <InputLabel
    className="lumx-progress-tracker-step__label"
    htmlFor="tab-3"
    theme="light"
  >
    Test Step Label
  </InputLabel>
</button>
`;

exports[`<ProgressTrackerStep> Snapshots and structure should render complete state 1`] = `
<button
  aria-controls="tab-panel-3"
  aria-selected={false}
  className="lumx-progress-tracker-step lumx-progress-tracker-step--is-clickable"
  id="tab-3"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={-1}
  type="button"
>
  <Icon
    className="lumx-progress-tracker-step__state"
    icon="M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
    size="s"
  />
  <InputLabel
    className="lumx-progress-tracker-step__label"
    htmlFor="tab-3"
    theme="light"
  >
    Test Step Label
  </InputLabel>
</button>
`;

exports[`<ProgressTrackerStep> Snapshots and structure should render erroneous state 1`] = `
<button
  aria-controls="tab-panel-3"
  aria-selected={false}
  className="lumx-progress-tracker-step lumx-progress-tracker-step--is-clickable"
  id="tab-3"
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  role="tab"
  tabIndex={-1}
  type="button"
>
  <Icon
    className="lumx-progress-tracker-step__state"
    icon="M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
    size="s"
  />
  <InputLabel
    className="lumx-progress-tracker-step__label"
    htmlFor="tab-3"
    theme="light"
  >
    Test Step Label
  </InputLabel>
</button>
`;
